{
 "cells": [
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "读取clinvar数据\n",
    "\n",
    "1. 过滤致病等级\n",
    "2. 过滤突变类型\n",
    "3. 过滤基因组版本\n",
    "4. 过滤MT染色体\n",
    "5. 过滤na的碱基\n",
    "7. 根据review status过滤可信度低的突变\n",
    "6. 给标签\n",
    "8. 写入variants_proessed.csv"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "import pandas as pd\n",
    "import os\n",
    "pd.options.display.max_columns = None\n",
    "\n",
    "\n",
    "clinvar_data_file = '/data/personal/zhenxy/clinvar/variant_summary.txt'\n",
    "if not os.path.exists(clinvar_data_file):\n",
    "    raise FileExistsError('The Clinvar Data File is not existed')\n",
    "raw_df = pd.read_csv(clinvar_data_file, delimiter='\\t')\n",
    "raw_df.head(3)\n",
    "raw_df = raw_df[(raw_df['ClinicalSignificance'] == 'Pathogenic') | (raw_df['ClinicalSignificance'] == 'Benign')]\n",
    "raw_df = raw_df[raw_df['Type']=='single nucleotide variant']\n",
    "raw_df = raw_df[raw_df['Assembly'] == 'GRCh38']\n",
    "raw_df = raw_df[raw_df['Chromosome']!='MT']\n",
    "raw_df = raw_df[raw_df['ReferenceAlleleVCF'] != 'na']\n",
    "review_status = [\n",
    "    'criteria provided, multiple submitters, no conflicts',\n",
    "    'reviewed by expert panel',\n",
    "    'practice guideline'\n",
    "    ]\n",
    "raw_df = raw_df[raw_df['ReviewStatus'].isin(review_status)]\n",
    "raw_df['label'] = -1\n",
    "raw_df['label'][raw_df['ClinicalSignificance'] == 'Pathogenic'] = 1\n",
    "raw_df['label'][raw_df['ClinicalSignificance'] != 'Pathogenic'] = 0\n",
    "processed_data_dir = '../data/clinvar/'\n",
    "if  not os.path.exists(processed_data_dir):\n",
    "    os.makedirs(processed_data_dir)\n",
    "file_name = 'processed_variant_summary.txt'\n",
    "\n",
    "raw_df.to_csv(os.path.join(processed_data_dir, file_name), sep='\\t', index=False)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "INFO: Pandarallel will run on 24 workers.\n",
      "INFO: Pandarallel will use Memory file system to transfer data between the main process and workers.\n",
      "genome sequences are loaded.\n"
     ]
    }
   ],
   "source": [
    "import pandas as pd\n",
    "from pandarallel import pandarallel\n",
    "\n",
    "pandarallel.initialize()\n",
    "\n",
    "chrs = [str(i) for i in range(1, 23)]\n",
    "chrs.extend(['X', 'Y'])\n",
    "chr_seqs = {}\n",
    "for chr in chrs:\n",
    "    with open(r'/data/personal/zhenxy/ucsc/hg38/chr{}.txt'.format(chr)) as f:\n",
    "        seq = f.readline()\n",
    "        chr_seqs[chr] = seq \n",
    "\n",
    "print(\"genome sequences are loaded.\")\n",
    "\n",
    "def get_seq(row, seq_len):\n",
    "    chr = row['Chromosome']\n",
    "    seq = chr_seqs[chr]\n",
    "    point_index = int(row['PositionVCF'])\n",
    "    ref = row['ReferenceAlleleVCF']\n",
    "    alt = row['AlternateAlleleVCF']\n",
    "    left = seq[point_index-1-seq_len:point_index-1]\n",
    "    right = seq[point_index: point_index+seq_len]\n",
    "    if len(left)<seq_len:\n",
    "        left = 'N' *(seq_len-len(left)) + left\n",
    "    if len(right)<seq_len:\n",
    "        right = right + 'N'*(seq_len-len(right))\n",
    "    t1 = ''.join([left, ref, right])\n",
    "    t2 = ''.join([left, alt, right])\n",
    "    return t1, t2"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>#AlleleID</th>\n",
       "      <th>Type</th>\n",
       "      <th>Name</th>\n",
       "      <th>GeneID</th>\n",
       "      <th>GeneSymbol</th>\n",
       "      <th>HGNC_ID</th>\n",
       "      <th>ClinicalSignificance</th>\n",
       "      <th>ClinSigSimple</th>\n",
       "      <th>LastEvaluated</th>\n",
       "      <th>RS# (dbSNP)</th>\n",
       "      <th>nsv/esv (dbVar)</th>\n",
       "      <th>RCVaccession</th>\n",
       "      <th>PhenotypeIDS</th>\n",
       "      <th>PhenotypeList</th>\n",
       "      <th>Origin</th>\n",
       "      <th>OriginSimple</th>\n",
       "      <th>Assembly</th>\n",
       "      <th>ChromosomeAccession</th>\n",
       "      <th>Chromosome</th>\n",
       "      <th>Start</th>\n",
       "      <th>Stop</th>\n",
       "      <th>ReferenceAllele</th>\n",
       "      <th>AlternateAllele</th>\n",
       "      <th>Cytogenetic</th>\n",
       "      <th>ReviewStatus</th>\n",
       "      <th>NumberSubmitters</th>\n",
       "      <th>Guidelines</th>\n",
       "      <th>TestedInGTR</th>\n",
       "      <th>OtherIDs</th>\n",
       "      <th>SubmitterCategories</th>\n",
       "      <th>VariationID</th>\n",
       "      <th>PositionVCF</th>\n",
       "      <th>ReferenceAlleleVCF</th>\n",
       "      <th>AlternateAlleleVCF</th>\n",
       "      <th>label</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>15044</td>\n",
       "      <td>single nucleotide variant</td>\n",
       "      <td>NM_017547.4(FOXRED1):c.694C&gt;T (p.Gln232Ter)</td>\n",
       "      <td>55572</td>\n",
       "      <td>FOXRED1</td>\n",
       "      <td>HGNC:26927</td>\n",
       "      <td>Pathogenic</td>\n",
       "      <td>1</td>\n",
       "      <td>Dec 30, 2019</td>\n",
       "      <td>267606829</td>\n",
       "      <td>-</td>\n",
       "      <td>RCV000000015|RCV000578659|RCV001194045</td>\n",
       "      <td>MONDO:MONDO:0032624,MedGen:C4748791,OMIM:61824...</td>\n",
       "      <td>Mitochondrial complex 1 deficiency, nuclear ty...</td>\n",
       "      <td>germline</td>\n",
       "      <td>germline</td>\n",
       "      <td>GRCh38</td>\n",
       "      <td>NC_000011.10</td>\n",
       "      <td>11</td>\n",
       "      <td>126275389</td>\n",
       "      <td>126275389</td>\n",
       "      <td>na</td>\n",
       "      <td>na</td>\n",
       "      <td>11q24.2</td>\n",
       "      <td>criteria provided, multiple submitters, no con...</td>\n",
       "      <td>3</td>\n",
       "      <td>-</td>\n",
       "      <td>N</td>\n",
       "      <td>ClinGen:CA113792,OMIM:613622.0001</td>\n",
       "      <td>3</td>\n",
       "      <td>5</td>\n",
       "      <td>126275389</td>\n",
       "      <td>C</td>\n",
       "      <td>T</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>15074</td>\n",
       "      <td>single nucleotide variant</td>\n",
       "      <td>NM_001201543.2(FAM161A):c.685C&gt;T (p.Arg229Ter)</td>\n",
       "      <td>84140</td>\n",
       "      <td>FAM161A</td>\n",
       "      <td>HGNC:25808</td>\n",
       "      <td>Pathogenic</td>\n",
       "      <td>1</td>\n",
       "      <td>Sep 15, 2021</td>\n",
       "      <td>267606794</td>\n",
       "      <td>-</td>\n",
       "      <td>RCV000000052|RCV001074032|RCV001090971|RCV0012...</td>\n",
       "      <td>MONDO:MONDO:0011630,MedGen:C1419614,OMIM:60606...</td>\n",
       "      <td>Retinitis pigmentosa 28|Retinal dystrophy|not ...</td>\n",
       "      <td>germline</td>\n",
       "      <td>germline</td>\n",
       "      <td>GRCh38</td>\n",
       "      <td>NC_000002.12</td>\n",
       "      <td>2</td>\n",
       "      <td>61840319</td>\n",
       "      <td>61840319</td>\n",
       "      <td>na</td>\n",
       "      <td>na</td>\n",
       "      <td>2p15</td>\n",
       "      <td>criteria provided, multiple submitters, no con...</td>\n",
       "      <td>6</td>\n",
       "      <td>-</td>\n",
       "      <td>N</td>\n",
       "      <td>OMIM:613596.0001,ClinGen:CA251355</td>\n",
       "      <td>3</td>\n",
       "      <td>35</td>\n",
       "      <td>61840319</td>\n",
       "      <td>G</td>\n",
       "      <td>A</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>15075</td>\n",
       "      <td>single nucleotide variant</td>\n",
       "      <td>NM_001201543.2(FAM161A):c.1309A&gt;T (p.Arg437Ter)</td>\n",
       "      <td>84140</td>\n",
       "      <td>FAM161A</td>\n",
       "      <td>HGNC:25808</td>\n",
       "      <td>Pathogenic</td>\n",
       "      <td>1</td>\n",
       "      <td>Sep 15, 2021</td>\n",
       "      <td>200691042</td>\n",
       "      <td>-</td>\n",
       "      <td>RCV000000053|RCV000153226|RCV000678572|RCV0007...</td>\n",
       "      <td>MONDO:MONDO:0011630,MedGen:C1419614,OMIM:60606...</td>\n",
       "      <td>Retinitis pigmentosa 28|not provided|Cone-rod ...</td>\n",
       "      <td>germline;inherited;unknown</td>\n",
       "      <td>germline</td>\n",
       "      <td>GRCh38</td>\n",
       "      <td>NC_000002.12</td>\n",
       "      <td>2</td>\n",
       "      <td>61839695</td>\n",
       "      <td>61839695</td>\n",
       "      <td>na</td>\n",
       "      <td>na</td>\n",
       "      <td>2p15</td>\n",
       "      <td>criteria provided, multiple submitters, no con...</td>\n",
       "      <td>19</td>\n",
       "      <td>-</td>\n",
       "      <td>N</td>\n",
       "      <td>ClinGen:CA233978,OMIM:613596.0002</td>\n",
       "      <td>3</td>\n",
       "      <td>36</td>\n",
       "      <td>61839695</td>\n",
       "      <td>T</td>\n",
       "      <td>A</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>15077</td>\n",
       "      <td>single nucleotide variant</td>\n",
       "      <td>NM_001201543.2(FAM161A):c.1567C&gt;T (p.Arg523Ter)</td>\n",
       "      <td>84140</td>\n",
       "      <td>FAM161A</td>\n",
       "      <td>HGNC:25808</td>\n",
       "      <td>Pathogenic</td>\n",
       "      <td>1</td>\n",
       "      <td>Apr 01, 2021</td>\n",
       "      <td>202193201</td>\n",
       "      <td>-</td>\n",
       "      <td>RCV000000055|RCV000787606|RCV000790648|RCV0010...</td>\n",
       "      <td>MONDO:MONDO:0011630,MedGen:C1419614,OMIM:60606...</td>\n",
       "      <td>Retinitis pigmentosa 28|Retinal dystrophy|not ...</td>\n",
       "      <td>germline;inherited;unknown</td>\n",
       "      <td>germline</td>\n",
       "      <td>GRCh38</td>\n",
       "      <td>NC_000002.12</td>\n",
       "      <td>2</td>\n",
       "      <td>61839437</td>\n",
       "      <td>61839437</td>\n",
       "      <td>na</td>\n",
       "      <td>na</td>\n",
       "      <td>2p15</td>\n",
       "      <td>criteria provided, multiple submitters, no con...</td>\n",
       "      <td>11</td>\n",
       "      <td>-</td>\n",
       "      <td>N</td>\n",
       "      <td>ClinGen:CA233972,OMIM:613596.0004</td>\n",
       "      <td>3</td>\n",
       "      <td>38</td>\n",
       "      <td>61839437</td>\n",
       "      <td>G</td>\n",
       "      <td>A</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>15099</td>\n",
       "      <td>single nucleotide variant</td>\n",
       "      <td>NM_006642.5(SDCCAG8):c.740+356C&gt;T</td>\n",
       "      <td>10806</td>\n",
       "      <td>SDCCAG8</td>\n",
       "      <td>HGNC:10671</td>\n",
       "      <td>Pathogenic</td>\n",
       "      <td>1</td>\n",
       "      <td>Jun 01, 2018</td>\n",
       "      <td>397515337</td>\n",
       "      <td>-</td>\n",
       "      <td>RCV000000077|RCV000760978</td>\n",
       "      <td>MONDO:MONDO:0014444,MedGen:C3889474,OMIM:61599...</td>\n",
       "      <td>Bardet-Biedl syndrome 16|Senior-Loken syndrome 7</td>\n",
       "      <td>germline</td>\n",
       "      <td>germline</td>\n",
       "      <td>GRCh38</td>\n",
       "      <td>NC_000001.11</td>\n",
       "      <td>1</td>\n",
       "      <td>243305133</td>\n",
       "      <td>243305133</td>\n",
       "      <td>na</td>\n",
       "      <td>na</td>\n",
       "      <td>1q43</td>\n",
       "      <td>criteria provided, multiple submitters, no con...</td>\n",
       "      <td>3</td>\n",
       "      <td>-</td>\n",
       "      <td>N</td>\n",
       "      <td>ClinGen:CA251367,OMIM:613524.0004</td>\n",
       "      <td>3</td>\n",
       "      <td>60</td>\n",
       "      <td>243305133</td>\n",
       "      <td>C</td>\n",
       "      <td>T</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "   #AlleleID                       Type  \\\n",
       "0      15044  single nucleotide variant   \n",
       "1      15074  single nucleotide variant   \n",
       "2      15075  single nucleotide variant   \n",
       "3      15077  single nucleotide variant   \n",
       "4      15099  single nucleotide variant   \n",
       "\n",
       "                                              Name  GeneID GeneSymbol  \\\n",
       "0      NM_017547.4(FOXRED1):c.694C>T (p.Gln232Ter)   55572    FOXRED1   \n",
       "1   NM_001201543.2(FAM161A):c.685C>T (p.Arg229Ter)   84140    FAM161A   \n",
       "2  NM_001201543.2(FAM161A):c.1309A>T (p.Arg437Ter)   84140    FAM161A   \n",
       "3  NM_001201543.2(FAM161A):c.1567C>T (p.Arg523Ter)   84140    FAM161A   \n",
       "4                NM_006642.5(SDCCAG8):c.740+356C>T   10806    SDCCAG8   \n",
       "\n",
       "      HGNC_ID ClinicalSignificance  ClinSigSimple LastEvaluated  RS# (dbSNP)  \\\n",
       "0  HGNC:26927           Pathogenic              1  Dec 30, 2019    267606829   \n",
       "1  HGNC:25808           Pathogenic              1  Sep 15, 2021    267606794   \n",
       "2  HGNC:25808           Pathogenic              1  Sep 15, 2021    200691042   \n",
       "3  HGNC:25808           Pathogenic              1  Apr 01, 2021    202193201   \n",
       "4  HGNC:10671           Pathogenic              1  Jun 01, 2018    397515337   \n",
       "\n",
       "  nsv/esv (dbVar)                                       RCVaccession  \\\n",
       "0               -             RCV000000015|RCV000578659|RCV001194045   \n",
       "1               -  RCV000000052|RCV001074032|RCV001090971|RCV0012...   \n",
       "2               -  RCV000000053|RCV000153226|RCV000678572|RCV0007...   \n",
       "3               -  RCV000000055|RCV000787606|RCV000790648|RCV0010...   \n",
       "4               -                          RCV000000077|RCV000760978   \n",
       "\n",
       "                                        PhenotypeIDS  \\\n",
       "0  MONDO:MONDO:0032624,MedGen:C4748791,OMIM:61824...   \n",
       "1  MONDO:MONDO:0011630,MedGen:C1419614,OMIM:60606...   \n",
       "2  MONDO:MONDO:0011630,MedGen:C1419614,OMIM:60606...   \n",
       "3  MONDO:MONDO:0011630,MedGen:C1419614,OMIM:60606...   \n",
       "4  MONDO:MONDO:0014444,MedGen:C3889474,OMIM:61599...   \n",
       "\n",
       "                                       PhenotypeList  \\\n",
       "0  Mitochondrial complex 1 deficiency, nuclear ty...   \n",
       "1  Retinitis pigmentosa 28|Retinal dystrophy|not ...   \n",
       "2  Retinitis pigmentosa 28|not provided|Cone-rod ...   \n",
       "3  Retinitis pigmentosa 28|Retinal dystrophy|not ...   \n",
       "4   Bardet-Biedl syndrome 16|Senior-Loken syndrome 7   \n",
       "\n",
       "                       Origin OriginSimple Assembly ChromosomeAccession  \\\n",
       "0                    germline     germline   GRCh38        NC_000011.10   \n",
       "1                    germline     germline   GRCh38        NC_000002.12   \n",
       "2  germline;inherited;unknown     germline   GRCh38        NC_000002.12   \n",
       "3  germline;inherited;unknown     germline   GRCh38        NC_000002.12   \n",
       "4                    germline     germline   GRCh38        NC_000001.11   \n",
       "\n",
       "  Chromosome      Start       Stop ReferenceAllele AlternateAllele  \\\n",
       "0         11  126275389  126275389              na              na   \n",
       "1          2   61840319   61840319              na              na   \n",
       "2          2   61839695   61839695              na              na   \n",
       "3          2   61839437   61839437              na              na   \n",
       "4          1  243305133  243305133              na              na   \n",
       "\n",
       "  Cytogenetic                                       ReviewStatus  \\\n",
       "0     11q24.2  criteria provided, multiple submitters, no con...   \n",
       "1        2p15  criteria provided, multiple submitters, no con...   \n",
       "2        2p15  criteria provided, multiple submitters, no con...   \n",
       "3        2p15  criteria provided, multiple submitters, no con...   \n",
       "4        1q43  criteria provided, multiple submitters, no con...   \n",
       "\n",
       "   NumberSubmitters Guidelines TestedInGTR                           OtherIDs  \\\n",
       "0                 3          -           N  ClinGen:CA113792,OMIM:613622.0001   \n",
       "1                 6          -           N  OMIM:613596.0001,ClinGen:CA251355   \n",
       "2                19          -           N  ClinGen:CA233978,OMIM:613596.0002   \n",
       "3                11          -           N  ClinGen:CA233972,OMIM:613596.0004   \n",
       "4                 3          -           N  ClinGen:CA251367,OMIM:613524.0004   \n",
       "\n",
       "   SubmitterCategories  VariationID  PositionVCF ReferenceAlleleVCF  \\\n",
       "0                    3            5    126275389                  C   \n",
       "1                    3           35     61840319                  G   \n",
       "2                    3           36     61839695                  T   \n",
       "3                    3           38     61839437                  G   \n",
       "4                    3           60    243305133                  C   \n",
       "\n",
       "  AlternateAlleleVCF  label  \n",
       "0                  T      1  \n",
       "1                  A      1  \n",
       "2                  A      1  \n",
       "3                  A      1  \n",
       "4                  T      1  "
      ]
     },
     "execution_count": 2,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "import pandas as pd\n",
    "pd.options.display.max_columns = None\n",
    "\n",
    "variant_df = pd.read_csv('../data/clinvar/processed_variant_summary.txt', sep='\\t')\n",
    "variant_df.head(5)\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "Benign        29927\n",
       "Pathogenic    10298\n",
       "Name: ClinicalSignificance, dtype: int64"
      ]
     },
     "execution_count": 5,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "variant_df['ClinicalSignificance'].value_counts()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "metadata": {},
   "outputs": [],
   "source": [
    "extra_len = 100\n",
    "res_file_path = \"../data/seq/{}/seq.csv\".format(2 * extra_len + 1)\n",
    "if not os.path.exists(os.path.dirname(res_file_path)):\n",
    "    os.makedirs(os.path.dirname(res_file_path))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "variant_df[['seq_ref', 'seq_alt']] = variant_df.parallel_apply(get_seq, result_type='expand', axis=1, seq_len=extra_len)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "variant_df[['ReferenceAlleleVCF', 'AlternateAlleleVCF', 'seq_ref', 'seq_alt', 'label']].to_csv(res_file_path, index=False)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 15,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>#AlleleID</th>\n",
       "      <th>Type</th>\n",
       "      <th>Name</th>\n",
       "      <th>GeneID</th>\n",
       "      <th>GeneSymbol</th>\n",
       "      <th>HGNC_ID</th>\n",
       "      <th>ClinicalSignificance</th>\n",
       "      <th>ClinSigSimple</th>\n",
       "      <th>LastEvaluated</th>\n",
       "      <th>RS# (dbSNP)</th>\n",
       "      <th>nsv/esv (dbVar)</th>\n",
       "      <th>RCVaccession</th>\n",
       "      <th>PhenotypeIDS</th>\n",
       "      <th>PhenotypeList</th>\n",
       "      <th>Origin</th>\n",
       "      <th>OriginSimple</th>\n",
       "      <th>Assembly</th>\n",
       "      <th>ChromosomeAccession</th>\n",
       "      <th>Chromosome</th>\n",
       "      <th>Start</th>\n",
       "      <th>Stop</th>\n",
       "      <th>ReferenceAllele</th>\n",
       "      <th>AlternateAllele</th>\n",
       "      <th>Cytogenetic</th>\n",
       "      <th>ReviewStatus</th>\n",
       "      <th>NumberSubmitters</th>\n",
       "      <th>Guidelines</th>\n",
       "      <th>TestedInGTR</th>\n",
       "      <th>OtherIDs</th>\n",
       "      <th>SubmitterCategories</th>\n",
       "      <th>VariationID</th>\n",
       "      <th>PositionVCF</th>\n",
       "      <th>ReferenceAlleleVCF</th>\n",
       "      <th>AlternateAlleleVCF</th>\n",
       "      <th>label</th>\n",
       "      <th>seq_ref</th>\n",
       "      <th>sequence</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>15044</td>\n",
       "      <td>single nucleotide variant</td>\n",
       "      <td>NM_017547.4(FOXRED1):c.694C&gt;T (p.Gln232Ter)</td>\n",
       "      <td>55572</td>\n",
       "      <td>FOXRED1</td>\n",
       "      <td>HGNC:26927</td>\n",
       "      <td>Pathogenic</td>\n",
       "      <td>1</td>\n",
       "      <td>Dec 30, 2019</td>\n",
       "      <td>267606829</td>\n",
       "      <td>-</td>\n",
       "      <td>RCV000000015|RCV000578659|RCV001194045</td>\n",
       "      <td>MONDO:MONDO:0032624,MedGen:C4748791,OMIM:61824...</td>\n",
       "      <td>Mitochondrial complex 1 deficiency, nuclear ty...</td>\n",
       "      <td>germline</td>\n",
       "      <td>germline</td>\n",
       "      <td>GRCh38</td>\n",
       "      <td>NC_000011.10</td>\n",
       "      <td>11</td>\n",
       "      <td>126275389</td>\n",
       "      <td>126275389</td>\n",
       "      <td>na</td>\n",
       "      <td>na</td>\n",
       "      <td>11q24.2</td>\n",
       "      <td>criteria provided, multiple submitters, no con...</td>\n",
       "      <td>3</td>\n",
       "      <td>-</td>\n",
       "      <td>N</td>\n",
       "      <td>ClinGen:CA113792,OMIM:613622.0001</td>\n",
       "      <td>3</td>\n",
       "      <td>5</td>\n",
       "      <td>126275389</td>\n",
       "      <td>C</td>\n",
       "      <td>T</td>\n",
       "      <td>1</td>\n",
       "      <td>CTCATCCCTCTTTGTGAGTTCTCTTTTTCTTATCACAGGGATGGAG...</td>\n",
       "      <td>CTCATCCCTCTTTGTGAGTTCTCTTTTTCTTATCACAGGGATGGAG...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>15074</td>\n",
       "      <td>single nucleotide variant</td>\n",
       "      <td>NM_001201543.2(FAM161A):c.685C&gt;T (p.Arg229Ter)</td>\n",
       "      <td>84140</td>\n",
       "      <td>FAM161A</td>\n",
       "      <td>HGNC:25808</td>\n",
       "      <td>Pathogenic</td>\n",
       "      <td>1</td>\n",
       "      <td>Sep 15, 2021</td>\n",
       "      <td>267606794</td>\n",
       "      <td>-</td>\n",
       "      <td>RCV000000052|RCV001074032|RCV001090971|RCV0012...</td>\n",
       "      <td>MONDO:MONDO:0011630,MedGen:C1419614,OMIM:60606...</td>\n",
       "      <td>Retinitis pigmentosa 28|Retinal dystrophy|not ...</td>\n",
       "      <td>germline</td>\n",
       "      <td>germline</td>\n",
       "      <td>GRCh38</td>\n",
       "      <td>NC_000002.12</td>\n",
       "      <td>2</td>\n",
       "      <td>61840319</td>\n",
       "      <td>61840319</td>\n",
       "      <td>na</td>\n",
       "      <td>na</td>\n",
       "      <td>2p15</td>\n",
       "      <td>criteria provided, multiple submitters, no con...</td>\n",
       "      <td>6</td>\n",
       "      <td>-</td>\n",
       "      <td>N</td>\n",
       "      <td>OMIM:613596.0001,ClinGen:CA251355</td>\n",
       "      <td>3</td>\n",
       "      <td>35</td>\n",
       "      <td>61840319</td>\n",
       "      <td>G</td>\n",
       "      <td>A</td>\n",
       "      <td>1</td>\n",
       "      <td>ATATCTGATTTAGATTTCATGGACTCTTCTTTTTTCTTCTGTTCTC...</td>\n",
       "      <td>ATATCTGATTTAGATTTCATGGACTCTTCTTTTTTCTTCTGTTCTC...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>15075</td>\n",
       "      <td>single nucleotide variant</td>\n",
       "      <td>NM_001201543.2(FAM161A):c.1309A&gt;T (p.Arg437Ter)</td>\n",
       "      <td>84140</td>\n",
       "      <td>FAM161A</td>\n",
       "      <td>HGNC:25808</td>\n",
       "      <td>Pathogenic</td>\n",
       "      <td>1</td>\n",
       "      <td>Sep 15, 2021</td>\n",
       "      <td>200691042</td>\n",
       "      <td>-</td>\n",
       "      <td>RCV000000053|RCV000153226|RCV000678572|RCV0007...</td>\n",
       "      <td>MONDO:MONDO:0011630,MedGen:C1419614,OMIM:60606...</td>\n",
       "      <td>Retinitis pigmentosa 28|not provided|Cone-rod ...</td>\n",
       "      <td>germline;inherited;unknown</td>\n",
       "      <td>germline</td>\n",
       "      <td>GRCh38</td>\n",
       "      <td>NC_000002.12</td>\n",
       "      <td>2</td>\n",
       "      <td>61839695</td>\n",
       "      <td>61839695</td>\n",
       "      <td>na</td>\n",
       "      <td>na</td>\n",
       "      <td>2p15</td>\n",
       "      <td>criteria provided, multiple submitters, no con...</td>\n",
       "      <td>19</td>\n",
       "      <td>-</td>\n",
       "      <td>N</td>\n",
       "      <td>ClinGen:CA233978,OMIM:613596.0002</td>\n",
       "      <td>3</td>\n",
       "      <td>36</td>\n",
       "      <td>61839695</td>\n",
       "      <td>T</td>\n",
       "      <td>A</td>\n",
       "      <td>1</td>\n",
       "      <td>TCTCTTTTAATAGATGCATGTGGAGATGCATGAAGATCAAATGGTT...</td>\n",
       "      <td>TCTCTTTTAATAGATGCATGTGGAGATGCATGAAGATCAAATGGTT...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>15077</td>\n",
       "      <td>single nucleotide variant</td>\n",
       "      <td>NM_001201543.2(FAM161A):c.1567C&gt;T (p.Arg523Ter)</td>\n",
       "      <td>84140</td>\n",
       "      <td>FAM161A</td>\n",
       "      <td>HGNC:25808</td>\n",
       "      <td>Pathogenic</td>\n",
       "      <td>1</td>\n",
       "      <td>Apr 01, 2021</td>\n",
       "      <td>202193201</td>\n",
       "      <td>-</td>\n",
       "      <td>RCV000000055|RCV000787606|RCV000790648|RCV0010...</td>\n",
       "      <td>MONDO:MONDO:0011630,MedGen:C1419614,OMIM:60606...</td>\n",
       "      <td>Retinitis pigmentosa 28|Retinal dystrophy|not ...</td>\n",
       "      <td>germline;inherited;unknown</td>\n",
       "      <td>germline</td>\n",
       "      <td>GRCh38</td>\n",
       "      <td>NC_000002.12</td>\n",
       "      <td>2</td>\n",
       "      <td>61839437</td>\n",
       "      <td>61839437</td>\n",
       "      <td>na</td>\n",
       "      <td>na</td>\n",
       "      <td>2p15</td>\n",
       "      <td>criteria provided, multiple submitters, no con...</td>\n",
       "      <td>11</td>\n",
       "      <td>-</td>\n",
       "      <td>N</td>\n",
       "      <td>ClinGen:CA233972,OMIM:613596.0004</td>\n",
       "      <td>3</td>\n",
       "      <td>38</td>\n",
       "      <td>61839437</td>\n",
       "      <td>G</td>\n",
       "      <td>A</td>\n",
       "      <td>1</td>\n",
       "      <td>AAATTTACCAACATAAACTCCACAGAGCTGCATACACTCATCTGGC...</td>\n",
       "      <td>AAATTTACCAACATAAACTCCACAGAGCTGCATACACTCATCTGGC...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>15099</td>\n",
       "      <td>single nucleotide variant</td>\n",
       "      <td>NM_006642.5(SDCCAG8):c.740+356C&gt;T</td>\n",
       "      <td>10806</td>\n",
       "      <td>SDCCAG8</td>\n",
       "      <td>HGNC:10671</td>\n",
       "      <td>Pathogenic</td>\n",
       "      <td>1</td>\n",
       "      <td>Jun 01, 2018</td>\n",
       "      <td>397515337</td>\n",
       "      <td>-</td>\n",
       "      <td>RCV000000077|RCV000760978</td>\n",
       "      <td>MONDO:MONDO:0014444,MedGen:C3889474,OMIM:61599...</td>\n",
       "      <td>Bardet-Biedl syndrome 16|Senior-Loken syndrome 7</td>\n",
       "      <td>germline</td>\n",
       "      <td>germline</td>\n",
       "      <td>GRCh38</td>\n",
       "      <td>NC_000001.11</td>\n",
       "      <td>1</td>\n",
       "      <td>243305133</td>\n",
       "      <td>243305133</td>\n",
       "      <td>na</td>\n",
       "      <td>na</td>\n",
       "      <td>1q43</td>\n",
       "      <td>criteria provided, multiple submitters, no con...</td>\n",
       "      <td>3</td>\n",
       "      <td>-</td>\n",
       "      <td>N</td>\n",
       "      <td>ClinGen:CA251367,OMIM:613524.0004</td>\n",
       "      <td>3</td>\n",
       "      <td>60</td>\n",
       "      <td>243305133</td>\n",
       "      <td>C</td>\n",
       "      <td>T</td>\n",
       "      <td>1</td>\n",
       "      <td>gcggtggctcacgcctgtaatcccagcactttgggaggctgaagcg...</td>\n",
       "      <td>GCGGTGGCTCACGCCTGTAATCCCAGCACTTTGGGAGGCTGAAGCG...</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "   #AlleleID                       Type  \\\n",
       "0      15044  single nucleotide variant   \n",
       "1      15074  single nucleotide variant   \n",
       "2      15075  single nucleotide variant   \n",
       "3      15077  single nucleotide variant   \n",
       "4      15099  single nucleotide variant   \n",
       "\n",
       "                                              Name  GeneID GeneSymbol  \\\n",
       "0      NM_017547.4(FOXRED1):c.694C>T (p.Gln232Ter)   55572    FOXRED1   \n",
       "1   NM_001201543.2(FAM161A):c.685C>T (p.Arg229Ter)   84140    FAM161A   \n",
       "2  NM_001201543.2(FAM161A):c.1309A>T (p.Arg437Ter)   84140    FAM161A   \n",
       "3  NM_001201543.2(FAM161A):c.1567C>T (p.Arg523Ter)   84140    FAM161A   \n",
       "4                NM_006642.5(SDCCAG8):c.740+356C>T   10806    SDCCAG8   \n",
       "\n",
       "      HGNC_ID ClinicalSignificance  ClinSigSimple LastEvaluated  RS# (dbSNP)  \\\n",
       "0  HGNC:26927           Pathogenic              1  Dec 30, 2019    267606829   \n",
       "1  HGNC:25808           Pathogenic              1  Sep 15, 2021    267606794   \n",
       "2  HGNC:25808           Pathogenic              1  Sep 15, 2021    200691042   \n",
       "3  HGNC:25808           Pathogenic              1  Apr 01, 2021    202193201   \n",
       "4  HGNC:10671           Pathogenic              1  Jun 01, 2018    397515337   \n",
       "\n",
       "  nsv/esv (dbVar)                                       RCVaccession  \\\n",
       "0               -             RCV000000015|RCV000578659|RCV001194045   \n",
       "1               -  RCV000000052|RCV001074032|RCV001090971|RCV0012...   \n",
       "2               -  RCV000000053|RCV000153226|RCV000678572|RCV0007...   \n",
       "3               -  RCV000000055|RCV000787606|RCV000790648|RCV0010...   \n",
       "4               -                          RCV000000077|RCV000760978   \n",
       "\n",
       "                                        PhenotypeIDS  \\\n",
       "0  MONDO:MONDO:0032624,MedGen:C4748791,OMIM:61824...   \n",
       "1  MONDO:MONDO:0011630,MedGen:C1419614,OMIM:60606...   \n",
       "2  MONDO:MONDO:0011630,MedGen:C1419614,OMIM:60606...   \n",
       "3  MONDO:MONDO:0011630,MedGen:C1419614,OMIM:60606...   \n",
       "4  MONDO:MONDO:0014444,MedGen:C3889474,OMIM:61599...   \n",
       "\n",
       "                                       PhenotypeList  \\\n",
       "0  Mitochondrial complex 1 deficiency, nuclear ty...   \n",
       "1  Retinitis pigmentosa 28|Retinal dystrophy|not ...   \n",
       "2  Retinitis pigmentosa 28|not provided|Cone-rod ...   \n",
       "3  Retinitis pigmentosa 28|Retinal dystrophy|not ...   \n",
       "4   Bardet-Biedl syndrome 16|Senior-Loken syndrome 7   \n",
       "\n",
       "                       Origin OriginSimple Assembly ChromosomeAccession  \\\n",
       "0                    germline     germline   GRCh38        NC_000011.10   \n",
       "1                    germline     germline   GRCh38        NC_000002.12   \n",
       "2  germline;inherited;unknown     germline   GRCh38        NC_000002.12   \n",
       "3  germline;inherited;unknown     germline   GRCh38        NC_000002.12   \n",
       "4                    germline     germline   GRCh38        NC_000001.11   \n",
       "\n",
       "  Chromosome      Start       Stop ReferenceAllele AlternateAllele  \\\n",
       "0         11  126275389  126275389              na              na   \n",
       "1          2   61840319   61840319              na              na   \n",
       "2          2   61839695   61839695              na              na   \n",
       "3          2   61839437   61839437              na              na   \n",
       "4          1  243305133  243305133              na              na   \n",
       "\n",
       "  Cytogenetic                                       ReviewStatus  \\\n",
       "0     11q24.2  criteria provided, multiple submitters, no con...   \n",
       "1        2p15  criteria provided, multiple submitters, no con...   \n",
       "2        2p15  criteria provided, multiple submitters, no con...   \n",
       "3        2p15  criteria provided, multiple submitters, no con...   \n",
       "4        1q43  criteria provided, multiple submitters, no con...   \n",
       "\n",
       "   NumberSubmitters Guidelines TestedInGTR                           OtherIDs  \\\n",
       "0                 3          -           N  ClinGen:CA113792,OMIM:613622.0001   \n",
       "1                 6          -           N  OMIM:613596.0001,ClinGen:CA251355   \n",
       "2                19          -           N  ClinGen:CA233978,OMIM:613596.0002   \n",
       "3                11          -           N  ClinGen:CA233972,OMIM:613596.0004   \n",
       "4                 3          -           N  ClinGen:CA251367,OMIM:613524.0004   \n",
       "\n",
       "   SubmitterCategories  VariationID  PositionVCF ReferenceAlleleVCF  \\\n",
       "0                    3            5    126275389                  C   \n",
       "1                    3           35     61840319                  G   \n",
       "2                    3           36     61839695                  T   \n",
       "3                    3           38     61839437                  G   \n",
       "4                    3           60    243305133                  C   \n",
       "\n",
       "  AlternateAlleleVCF  label  \\\n",
       "0                  T      1   \n",
       "1                  A      1   \n",
       "2                  A      1   \n",
       "3                  A      1   \n",
       "4                  T      1   \n",
       "\n",
       "                                             seq_ref  \\\n",
       "0  CTCATCCCTCTTTGTGAGTTCTCTTTTTCTTATCACAGGGATGGAG...   \n",
       "1  ATATCTGATTTAGATTTCATGGACTCTTCTTTTTTCTTCTGTTCTC...   \n",
       "2  TCTCTTTTAATAGATGCATGTGGAGATGCATGAAGATCAAATGGTT...   \n",
       "3  AAATTTACCAACATAAACTCCACAGAGCTGCATACACTCATCTGGC...   \n",
       "4  gcggtggctcacgcctgtaatcccagcactttgggaggctgaagcg...   \n",
       "\n",
       "                                            sequence  \n",
       "0  CTCATCCCTCTTTGTGAGTTCTCTTTTTCTTATCACAGGGATGGAG...  \n",
       "1  ATATCTGATTTAGATTTCATGGACTCTTCTTTTTTCTTCTGTTCTC...  \n",
       "2  TCTCTTTTAATAGATGCATGTGGAGATGCATGAAGATCAAATGGTT...  \n",
       "3  AAATTTACCAACATAAACTCCACAGAGCTGCATACACTCATCTGGC...  \n",
       "4  GCGGTGGCTCACGCCTGTAATCCCAGCACTTTGGGAGGCTGAAGCG...  "
      ]
     },
     "execution_count": 15,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "variant_df.head(5)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 14,
   "metadata": {},
   "outputs": [],
   "source": [
    "\"\"\"\n",
    "突变后的序列\n",
    "\"\"\"\n",
    "variant_df.rename(columns={'seq_alt' : 'sequence'}, inplace=True)\n",
    "variant_df['sequence'] = variant_df['sequence'].apply(lambda x: x.upper())\n",
    "variant_df[['sequence', 'label']].to_csv(res_file_path, index=False)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 40,
   "metadata": {},
   "outputs": [],
   "source": [
    "\"\"\"\n",
    "split train test dataset\n",
    "\"\"\"\n",
    "\n",
    "import pandas as pd\n",
    "from sklearn.model_selection import train_test_split\n",
    "import numpy as np\n",
    "import os\n",
    "\n",
    "data_dir = '../data/seq/201/5/'\n",
    "data_file_path = '../data/seq/201/5/kmers5.csv'\n",
    "\n",
    "print(\"processing {}\".format(data_file_path))\n",
    "data = pd.read_csv(data_file_path, delimiter='\\t')\n",
    "data = data.values\n",
    "seqs, label = data[:, 0], data[:, 1]\n",
    "\n",
    "train_data, val_test_data, train_label, val_test_label = train_test_split(seqs, label, train_size=0.7, random_state=0, stratify=label) \n",
    "print(train_data.shape, val_test_data.shape, train_label.shape, val_test_label.shape)\n",
    "\n",
    "val_data, test_data, val_label, test_label = train_test_split(val_test_data, val_test_label, test_size=0.5, random_state=0, stratify=val_test_label)\n",
    "print(val_data.shape, val_label.shape, test_data.shape, test_label.shape)\n",
    "\n",
    "train_dataset = np.column_stack((train_data, train_label))\n",
    "val_dataset = np.column_stack((val_data, val_label))\n",
    "test_dataset = np.column_stack((test_data, test_label))\n",
    "\n",
    "train_dataset = np.insert(train_dataset, 0, ['sequence', 'label'], axis=0)\n",
    "val_dataset = np.insert(val_dataset, 0, ['sequence', 'label'], axis=0)\n",
    "test_dataset = np.insert(test_dataset, 0, ['sequence', 'label'], axis=0)\n",
    "\n",
    "\n",
    "train_dir = os.path.join(data_dir, 'train')\n",
    "if not os.path.exists(train_dir):\n",
    "    os.makedirs(train_dir)\n",
    "\n",
    "\n",
    "np.savetxt(os.path.join(train_dir, 'train.tsv'), train_dataset, delimiter='\\t', fmt='%s')\n",
    "np.savetxt(os.path.join(train_dir, 'dev.tsv'), val_dataset, delimiter='\\t', fmt='%s')\n",
    "\n",
    "test_dir = os.path.join(data_dir, 'test')\n",
    "if not os.path.exists(test_dir):\n",
    "    os.makedirs(test_dir)\n",
    "\n",
    "np.savetxt(os.path.join(test_dir, 'dev.tsv'), test_dataset, delimiter='\\t', fmt='%s')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 37,
   "metadata": {},
   "outputs": [],
   "source": [
    "from sklearn.model_selection import StratifiedShuffleSplit"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 14,
   "metadata": {},
   "outputs": [],
   "source": [
    "import numpy as np\n",
    "\n",
    "result  = np.load('../result/mutation/pred_results.npy')\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 15,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "(6034,)"
      ]
     },
     "execution_count": 15,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "result.shape"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 18,
   "metadata": {},
   "outputs": [],
   "source": [
    "import pandas as pd\n",
    "\n",
    "\n",
    "test_data = pd.read_csv('../data/seq/201/6/test/dev.tsv', sep='\\t')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 25,
   "metadata": {},
   "outputs": [],
   "source": [
    "truths = test_data['label'].to_numpy()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 29,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "Counter({0: 4489, 1: 1545})"
      ]
     },
     "execution_count": 29,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "from collections import Counter\n",
    "\n",
    "counter = Counter(truths)\n",
    "counter"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 33,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "              precision    recall  f1-score   support\n",
      "\n",
      "           0       0.83      0.90      0.87      4489\n",
      "           1       0.62      0.48      0.54      1545\n",
      "\n",
      "    accuracy                           0.79      6034\n",
      "   macro avg       0.73      0.69      0.70      6034\n",
      "weighted avg       0.78      0.79      0.78      6034\n",
      "\n"
     ]
    }
   ],
   "source": [
    "from sklearn.metrics import classification_report,confusion_matrix\n",
    "\n",
    "preds = np.where(result<=0.5, 0, 1)\n",
    "print(classification_report(truths, preds))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 34,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "array([[4045,  444],\n",
       "       [ 806,  739]])"
      ]
     },
     "execution_count": 34,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "confusion_matrix(truths, preds)"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3.6.15 ('dnabert')",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.6.15"
  },
  "orig_nbformat": 4,
  "vscode": {
   "interpreter": {
    "hash": "be6571a2952341348c05b61e867709023275420d04272653f792c5bd62ebb5be"
   }
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
